app/models/promo_registration.rb

```diff
@@ -4,26 +4,29 @@ class PromoRegistration < ApplicationRecord
   # A promo registration can belong to a channel,
   # a publisher, or be unattached.  Unattached codes
   # are created by admins.
-
-  # Constants
   CHANNEL = "channel".freeze
   OWNER = "owner".freeze
   UNATTACHED ="unattached".freeze
   KINDS = [CHANNEL, OWNER, UNATTACHED].freeze
 
+  # Event types
   RETRIEVALS = "retrievals" # Aliased as 'Downloads'
   FIRST_RUNS = "first_runs" # Aliased as 'Installs'
   FINALIZED = "finalized" # Aliased as 'Confirmed'
 
+  # Reporting intervals
   DAILY = "daily"
   WEEKLY = "weekly"
   MONTHLY = "monthly"
   RUNNING_TOTAL = "running_total"
   
   belongs_to :channel, validate: true, autosave: true
   belongs_to :promo_campaign
+  belongs_to :publisher, dependent: :destroy
 
-  validates :channel_id, presence: true, if: -> { kind == CHANNEL}
+  validates :channel_id, presence: true, if: -> { kind == CHANNEL }
+  validates :publisher_id, presence: true, if: -> { kind != UNATTACHED }
```

Comment:
Might read a little better if we used `unless`
```suggestion
  validates :publisher_id, presence: true, unless: -> { kind == UNATTACHED }
```